File: src_local/utils/stability_ledger.py
"""
SISTEMA DE PERSONAS AGENTES - MEMÓRIA DE ESTABILIDADE
Módulo: Livro de Estabilidade (StabilityLedger)
Função: Rastrear o histórico de falhas e sucessos de auto-cura.
Soberania: CORE-MEMORY.
"""
import logging
import time
from pathlib import Path

logger = logging.getLogger(__name__)

class StabilityLedger:
    """
    🏥 Livro de Estabilidade PhD.
    A memória de longo prazo do Orquestrador, responsável por catalogar recorrências
    de falhas, validar sucessos de auto-cura e manter a persistência da consciência sistêmica.
    """
    
    def __init__(self, project_root):
        self.project_root = Path(project_root)
        storage_path = self.project_root / ".gemini" / "stability_ledger.json"
        
        # Agente de Suporte à Persistência
        from src_local.agents.Support.memory_persistence import MemoryPersistence
        self.persistence = MemoryPersistence(storage_path)
        self.ledger = self.persistence.load_ledger()

    def update(self, audit_results, context_map=None):
        """
        📈 Sincroniza os achados da auditoria com a memória persistente.
        Identifica novas feridas e confirma a cicatrização de falhas anteriores.
        """
        current_error_files = set()
        context_map = context_map or {}
        is_internal = "Personas_Agentes" in str(self.project_root)

        try:
            # Limpeza preventiva para projetos externos
            if not is_internal and ".agent/skills" in self.ledger:
                del self.ledger[".agent/skills"]

            for issue in audit_results:
                file = str(issue.get('file', 'N/A')).replace("\\", "/") if isinstance(issue, dict) else "Strategic/DNA"
                current_error_files.add(file)
                self._update_file_status(file, context_map)

            # Detecção de CURA: Arquivos que estavam instáveis mas não reportaram erros
            self._detect_healed_files(current_error_files)
            
            self.persistence.save_ledger(self.ledger)
            return self.ledger
        except Exception as e:
            logger.error(f"🚨 Falha ao atualizar livro de estabilidade: {e}", exc_info=True)
            return self.ledger

    def _update_file_status(self, file, context_map):
        """
        🧬 Atualiza os metadados de instabilidade de um componente.
        Garante que Gold Standards sejam preservados e arquivos instáveis sejam marcados.
        """
        # Veto de Intencionalidade: Referências (Gold Standard) são imunes a instabilidade
        if context_map.get(file, {}).get("is_gold_standard"):
            if file not in self.ledger:
                self.ledger[file] = {"status": "REFERENCE", "history": ["Identificado como Gold Standard"]}
            return

        if file not in self.ledger:
            self.ledger[file] = {"occurrences": 0, "history": [], "status": "UNSTABLE"}
        
        self.ledger[file]["occurrences"] += 1
        self.ledger[file]["history"].append(time.strftime('%Y-%m-%d %H:%M:%S'))
        self.ledger[file]["status"] = "UNSTABLE"

    def _detect_healed_files(self, current_errors):
        """
        ✨ Compara o estado atual com a memória para confirmar curas atômicas.
        Reseta o contador de ocorrências ao validar a integridade restaurada.
        """
        for file in list(self.ledger.keys()):
            if file == "Strategic/DNA" or self.ledger[file].get("status") == "REFERENCE":
                continue
                
            if file not in current_errors and self.ledger[file].get("status") != "HEALED":
                logger.info(f"✨ [Memória] Cura confirmada: {file}")
                self.ledger[file]["status"] = "HEALED"
                self.ledger[file]["occurrences"] = 0

    def get_file_data(self, file_path):
        """Recupera metadados históricos de um arquivo."""
        # Agente de Suporte à Persistência
        from src_local.agents.Support.memory_persistence import MemoryPersistence
        storage_path = self.project_root / ".gemini" / "stability_ledger.json"
        persistence = MemoryPersistence(storage_path)
        return persistence.get_file_metadata(self.ledger, file_path)

File: src_local/utils/stability_ledger.ts
import winston from "winston";
import { Path } from "../core/path_utils.ts";
import { MemoryPersistence } from "../agents/Support/Core/memory_persistence.ts";

const logger = winston.child({ module: "StabilityLedger" });

/**
 * 🏥 Livro de Estabilidade PhD (Bun Version).
 */
export class StabilityLedger {
    projectRoot: Path;
    persistence: MemoryPersistence;
    ledger: Record<string, any>;

    constructor(projectRoot: string) {
        this.projectRoot = new Path(projectRoot);
        const storagePath = this.projectRoot.join(".gemini", "stability_ledger.json");

        this.persistence = new MemoryPersistence(storagePath);
        this.ledger = this.persistence.loadLedger();
    }

    async sync() {
        logger.info("💾 [Ledger] Sincronizando livro de estabilidade...");
        this.persistence.saveLedger(this.ledger);
    }

    update(auditResults: any[], contextMap: Record<string, any> = {}): Record<string, any> {
        try {
            this.performMaintenance();
            const currentErrorFiles = this.processAuditResults(auditResults, contextMap);
            this.detectHealedFiles(currentErrorFiles);

            this.persistence.saveLedger(this.ledger);
            return this.ledger;
        } catch (e: any) {
            logger.error(`🚨 Falha ao atualizar livro de estabilidade: ${e.message}`);
            return this.ledger;
        }
    }

    private performMaintenance() {
        const isInternal = this.projectRoot.toString().includes("Personas_Agentes");
        if (!isInternal && this.ledger[".agent/skills"]) {
            delete this.ledger[".agent/skills"];
        }
    }

    private processAuditResults(results: any[], contextMap: any): Set<string> {
        const currentErrorFiles = new Set<string>();
        results.forEach(issue => {
            const file = this.extractFileName(issue);
            currentErrorFiles.add(file);
            this.updateFileStatus(file, contextMap);
        });
        return currentErrorFiles;
    }

    private extractFileName(issue: any): string {
        if (typeof issue === 'object' && issue !== null && issue.file) {
            return String(issue.file).replace(/\\/g, "/");
        }
        return "Strategic/DNA";
    }

    private updateFileStatus(file: string, contextMap: Record<string, any>) {
        if (this.tryMarkReference(file, contextMap)) return;

        const entry = this.ledger[file] || { occurrences: 0, history: [], status: "UNSTABLE" };
        entry.occurrences += 1;
        entry.history.push(new Date().toISOString());

        // Only set status to UNSTABLE if it's not a REFERENCE
        if (entry.status !== "REFERENCE") {
            entry.status = "UNSTABLE";
        }

        this.ledger[file] = entry;
    }

    private tryMarkReference(file: string, contextMap: Record<string, any>): boolean {
        if (contextMap[file]?.is_gold_standard) {
            this.ledger[file] ||= { status: "REFERENCE", history: ["Identificado como Gold Standard"] };
            return true;
        }
        return false;
    }

    private detectHealedFiles(currentErrors: Set<string>) {
        Object.keys(this.ledger).forEach(file => {
            if (this.isHealable(file, currentErrors)) {
                this.markAsHealed(file);
            }
        });
    }

    private markAsHealed(file: string) {
        logger.info(`✨ [Memória] Cura confirmada: ${file}`);
        this.ledger[file].status = "HEALED";
        this.ledger[file].occurrences = 0;
    }

    private isHealable(file: string, currentErrors: Set<string>): boolean {
        const entry = this.ledger[file];
        if (file === "Strategic/DNA" || entry.status === "REFERENCE") return false;
        return !currentErrors.has(file) && entry.status !== "HEALED";
    }

    clear() {
        this.ledger = this.filterLedgerByStatus("REFERENCE");
        this.persistence.saveLedger(this.ledger);
        logger.info("🧹 [StabilityLedger] Memória de instabilidade resetada.");
    }

    private filterLedgerByStatus(status: string): Record<string, any> {
        return Object.fromEntries(
            Object.entries(this.ledger).filter(([_, v]) => v.status === status)
        );
    }

    getFileData(filePath: string): any {
        return this.persistence.getFileMetadata(this.ledger, filePath);
    }

    registerDisparity(disparity: { source: string, target: string, discrepancies: string[], severity: string }) {
        const file = disparity.source.replace(/\\/g, "/");
        const entry = this.ledger[file] || { occurrences: 0, history: [], status: "UNSTABLE" };
        
        entry.status = "DISPARITY";
        entry.occurrences += 1;
        entry.history.push(`[PARITY ERROR] Disparidade com ${disparity.target}: ${disparity.discrepancies.join("; ")}`);
        entry.meta = { ...entry.meta, lastSiblingDisparity: disparity.target, severity: disparity.severity };

        this.ledger[file] = entry;
        this.persistence.saveLedger(this.ledger);
        logger.warn(`⚖️ [Ledger] Disparidade registrada para ${file}`);
    }
}
